std/range/package.d

```diff
@@ -5353,284 +5353,387 @@ pure @safe nothrow unittest
-    alias Value = CommonType!(Unqual!B, Unqual!E);
-    alias StepType = Unqual!S;
+/// Ditto
+auto iota(B, E)(B begin, E end)
+{
+    static if (!is(B == Unqual!B) || !is(E == Unqual!E))
+    {
+        return iota(Unqual!B(begin), Unqual!E(end));
+    }
+    else static if (is(typeof(begin + B(1))))
+    {
+        return iota(begin, end, B(1));
+    }
+    else static if (is(typeof(begin + 1)))
+    {
+        return iota(begin, end, 1);
+    }
+    else static if (is(typeof(++begin)))
+    {
+        return IotaResult!(B, E)(begin, end);
+    }
+    else
+    {
+        static assert(false, "iota does not support the types ("~B.stringof~", "~E.stringof~").");
+    }
+}
 
-    assert(step != 0 || begin == end);
+/// Ditto
+auto iota(E)(E end)
+if (is(typeof(E(0))))
+{
+    static if (is(E == Unqual!E))
+        return iota(E(0), end);
+    else
+        return iota(Unqual!E(end));
+}
 
-    static struct Result
+/*
+ * Overload for anything supporting regular arithmetic.
+ */
+struct IotaResult(B, E, S)
+if (!anySatisfy!(isFloatingPoint, B, E, S) && is(typeof(B.init + S.init)))
+{
+    static if (is(CommonType!(B, E, S)) && !is(CommonType!(B, E, S) == void))
+        alias ValueType = CommonType!(B, E, S);
+    else static if (is(typeof(B.init + E.init + S.init)))
+        alias ValueType = typeof(B.init + E.init + S.init);
+    else static if (is(typeof(B.init + S.init)))
+        alias ValueType = typeof(B.init + S.init);
+    else
+        static assert(false, "iota does not support types that cannot be added or incremented.");
+
+    ValueType begin;
+    ValueType end;
+    S step;
+
+    private auto unsigned(T)(T value)
     {
-        private Value current, last;
-        private StepType step; // by convention, 0 if range is empty
+        static if (is(Unsigned!T)) return cast(Unsigned!T) value;
+        else return value;
+    }
 
-        this(Value current, Value pastLast, StepType step)
-        {
-            if (current < pastLast && step > 0)
-            {
-                // Iterating upward
-                assert(unsigned((pastLast - current) / step) <= size_t.max);
-                // Cast below can't fail because current < pastLast
-                this.last = cast(Value) (pastLast - 1);
-                this.last -= unsigned(this.last - current) % step;
-            }
-            else if (current > pastLast && step < 0)
-            {
-                // Iterating downward
-                assert(unsigned((current - pastLast) / (0 - step)) <= size_t.max);
-                // Cast below can't fail because current > pastLast
-                this.last = cast(Value) (pastLast + 1);
-                this.last += unsigned(current - this.last) % (0 - step);
-            }
-            else
-            {
-                // Initialize an empty range
-                this.step = 0;
-                return;
-            }
-            this.step = step;
-            this.current = current;
-        }
+    this()(B b, E e, S s)
```

Comment:
What's the reason for templating the constructor? `IotaResult` is already a template.